File: week08/15/2_Tue_MyCalendar.py
class MyCalendarTwo:

    def __init__(self):
        self.booked = []
        self.overlapped = []

    def book(self, start: int, end: int) -> bool:
        for o_start, o_end in self.overlapped:
            if max(start, o_start) < min(end, o_end):
                return False
        
        for b_start, b_end in self.booked:
            if max(start, b_start) < min(end, b_end):
                self.overlapped.append((max(start, b_start), min(end, b_end)))
                
        self.booked.append((start, end))
        
        return True
    
# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)

File: week08/15/2_Tue_MyCalendar.cpp
#include <vector>
#include <algorithm>

using namespace std;

class MyCalendarTwo {
public:
    vector<pair<int, int>> booked;
    vector<pair<int, int>> overlap;
    
    MyCalendarTwo() {        
    }
    
    bool book(int start, int end) {
        for (auto& o : overlap) {
            if (max(start, o.first) < min(end, o.second)) return false;
        }
        for (auto& b : booked) {
            if (max(start, b.first) < min(end, b.second)) 
                overlap.push_back({max(start, b.first), min(end, b.second)});
        }
        booked.push_back({start, end});
        return true;
    }
};

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo* obj = new MyCalendarTwo();
 * bool param_1 = obj->book(start,end);
 */